File: apps/api/hunter_api/schemas/radar_coverage.py
"""``GET /api/v1/radar/coverage`` — an honest snapshot of how much Radar
exists yet (T3.46's quant study, ``.claude/state/notes-T3.46.md``): coverage
of the monitored universe, baseline maturity, detector health and the
highest score ever recorded. Additive to the T2.6 radar contract — nothing
on ``schemas/radar.py`` changes.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel

from hunter_api.schemas.radar import DecimalStr
from hunter_core.domain.enums import AnomalyType

__all__ = ["RadarCoverageOut", "RadarDetectorOut"]


class RadarDetectorOut(BaseModel):
    """One of the twelve ``AnomalyType`` members against what it has actually
    produced in the last 31 days, plus why it is silent when it is.
    """

    type: AnomalyType
    rows_31d: int
    disarmed_reason: str | None = None
    """The scanner's own declared reason (``hb:scanner:*``'s
    ``detectors_disarmed`` field, ``services/scanner-worker
    /hunter_scanner_worker/health.py``) — ``None`` both when the detector is
    producing AND when it is silent with no reason declared. The two must
    stay tellable apart on screen (``rows_31d`` does that): a detector with
    ``rows_31d == 0`` and ``disarmed_reason is None`` is the defect T3.46
    flagged (``ORDERBOOK_IMBALANCE``, ``OPEN_INTEREST_SPIKE``,
    ``TRADE_VELOCITY_SPIKE`` today) — silence with no reason declared,
    exactly what this project's own honesty rule forbids."""


class RadarCoverageOut(BaseModel):
    markets_monitored: int
    markets_with_anomaly: int
    baselines_usable: int
    baselines_under_construction: int
    bootstrap_pointer: str | None
    """The scanner's own ``baselines_state`` heartbeat sentence (e.g.
    ``"bootstrapping 1000FLOKIUSDT (4/200)"``) — ``None`` only when no
    ``hb:scanner:*`` heartbeat could be read at all, never an invented
    "idle"."""
    baseline_gate_v2_pct: DecimalStr | None
    """% of ``feature_baselines`` rows whose ``distinct_days``/``sample_size``
    clear the *active* ``opportunity_weights.weights["baseline_gate"]`` —
    ``None`` when there is no active weight vector or no baseline row exists
    yet, never a fabricated ``0``."""
    detectors: list[RadarDetectorOut]
    max_score_ever: DecimalStr | None
    first_anomaly_at: datetime | None
    as_of: datetime

File: apps/web/tests/fixtures/radar-coverage.ts
/**
 * Fixture mirroring the real `GET /api/v1/radar/coverage` contract
 * (`apps/api/hunter_api/schemas/radar_coverage.py`, T3.46) -- never an
 * invented shape, per CLAUDE.md. Defaults are the actual T3.46 numbers
 * (`.claude/state/notes-T3.46.md`): 25/217 markets, ~8% baselines usable,
 * max score ever 38.33, four of twelve detectors producing.
 */
import type { RadarCoverageOut, RadarDetectorOut } from "@/lib/api/radar-coverage-types";

export function makeRadarDetector(overrides: Partial<RadarDetectorOut> = {}): RadarDetectorOut {
  return {
    type: "VOLUME_SPIKE",
    rows_31d: 575,
    disarmed_reason: null,
    ...overrides,
  };
}

export function makeRadarCoverage(overrides: Partial<RadarCoverageOut> = {}): RadarCoverageOut {
  return {
    markets_monitored: 217,
    markets_with_anomaly: 25,
    baselines_usable: 9029,
    baselines_under_construction: 102597,
    bootstrap_pointer: "bootstrapping 1000FLOKIUSDT (4/200)",
    baseline_gate_v2_pct: "4.28",
    detectors: [
      makeRadarDetector({ type: "VOLUME_SPIKE", rows_31d: 575 }),
      makeRadarDetector({ type: "MOMENTUM_SHIFT", rows_31d: 335 }),
      makeRadarDetector({ type: "PRICE_ACCELERATION", rows_31d: 329 }),
      makeRadarDetector({ type: "VOLATILITY_EXPANSION", rows_31d: 48 }),
      makeRadarDetector({
        type: "CROSS_EXCHANGE_DIVERGENCE",
        rows_31d: 0,
        disarmed_reason: "single_exchange_until_m1b",
      }),
      makeRadarDetector({ type: "FUNDING_ANOMALY", rows_31d: 0, disarmed_reason: "funding_unavailable" }),
      makeRadarDetector({
        type: "LIQUIDATION_CLUSTER",
        rows_31d: 0,
        disarmed_reason: "feature_not_implemented",
      }),
      makeRadarDetector({ type: "ORDERBOOK_IMBALANCE", rows_31d: 0, disarmed_reason: null }),
      makeRadarDetector({ type: "OPEN_INTEREST_SPIKE", rows_31d: 0, disarmed_reason: null }),
      makeRadarDetector({ type: "TRADE_VELOCITY_SPIKE", rows_31d: 0, disarmed_reason: null }),
      makeRadarDetector({ type: "SOCIAL_SPIKE", rows_31d: 0, disarmed_reason: null }),
      makeRadarDetector({ type: "WHALE_ACTIVITY", rows_31d: 0, disarmed_reason: null }),
    ],
    max_score_ever: "38.33",
    first_anomaly_at: "2026-09-07T02:24:01Z",
    as_of: "2026-09-08T22:37:00Z",
    ...overrides,
  };
}
